### PR TITLE
observability: add SessionPool span annotations

### PR DESCRIPTION
**Pull request by odeke-em**
This change adds span annotations on the active span which could either be the noopSpan or the top-level active one, for whenever we progressively add more spans.

Built from #2087
Updates #2079

### CODE REVIEW

**Comment by odeke-em** [comment]
## Exhibit
On running this code per instructions of sample https://gist.github.com/odeke-em/35fb0e58b3a73dca59730b688817be1b, we get back
Screenshot 2024-08-30 at 6 47 59 AM
Screenshot 2024-08-30 at 6 48 29 AM
Screenshot 2024-08-30 at 7 07 24 AM

File: src/session-pool.ts [hunk 1]
```diff
@@ -1048,6 +1112,9 @@ export class SessionPool extends EventEmitter implements SessionPoolInterface {
    * @param {Session} session The session object.
    */
   _release(session: Session): void {
+    const span = getActiveOrNoopSpan();
+    span.addEvent('releasing session back to the pool', {id: session.id});
```
Comment (surbhigarg92) [inline src/session-pool.ts:1116]: Remove this span. Removed session is enough.

File: src/session-pool.ts [hunk 2]
```diff
@@ -736,6 +755,8 @@ export class SessionPool extends EventEmitter implements SessionPoolInterface {
     while (amount > 0) {
       let sessions: Session[] | null = null;
 
+      span.addEvent(`trying to create ${amount} sessions`);
```
Comment (surbhigarg92) [inline src/session-pool.ts:758]: nit: rename to `Creating ${amount} sessions`

File: src/session-pool.ts [hunk 3]
```diff
@@ -919,9 +963,11 @@ export class SessionPool extends EventEmitter implements SessionPoolInterface {
     // Only create a new session if there are more waiters than sessions already
     // being created. The current requester will be waiter number _numWaiters+1.
     if (!this.isFull && this.totalPending <= this.totalWaiters) {
+      span.addEvent('pool is not full and more waiters than sessions');
```
Comment (surbhigarg92) [inline src/session-pool.ts:966]: Remove this, "Creating session" should be enough

File: src/session-pool.ts [hunk 4]
```diff
@@ -990,6 +1044,10 @@ export class SessionPool extends EventEmitter implements SessionPoolInterface {
    * @returns {Promise}
    */
   async _ping(session: Session): Promise<void> {
+    // NOTE: Please do not trace Ping as it gets quite spammy
```
Comment (surbhigarg92) [inline src/session-pool.ts:1047]: How are we making sure this does not happen. ?

File: src/session-pool.ts [hunk 5]
```diff
@@ -1030,10 +1088,16 @@ export class SessionPool extends EventEmitter implements SessionPoolInterface {
    * @param {object} options The transaction options.
    */
   _prepareTransaction(session: Session): void {
+    const span = getActiveOrNoopSpan();
+    span.addEvent('creating transaction for session');
```
Comment (surbhigarg92) [inline src/session-pool.ts:1092]: this actually does not create a transaction. Its just creates a transaction object. transaction.id will always be null here. 

Remove both the events here

Comment (surbhigarg92) [reply src/session-pool.ts:1092]: In database class we have `transaction.begin` calls, there we should add these events. 

We can simply add these in transaction.begin() method.

File: package.json [hunk 6]
```diff
@@ -60,7 +60,6 @@
     "@google-cloud/promisify": "^4.0.0",
     "@grpc/proto-loader": "^0.7.0",
     "@opentelemetry/api": "^1.9.0",
-    "@opentelemetry/context-async-hooks": "^1.25.1",
```
[the remark this answers is not in the document]
Comment (surbhigarg92) [reply package.json:63]: What does this package do, does it add any extra process which runs within app ?

File: src/session-pool.ts [hunk 7]
```diff
@@ -865,21 +892,35 @@ export class SessionPool extends EventEmitter implements SessionPoolInterface {
    * @returns {Promise<Session>}
    */
   async _getSession(startTime: number): Promise<Session> {
+    const span = getActiveOrNoopSpan();
+    span.addEvent('Attempting to get session');
+
     if (this._hasSessionUsableFor()) {
-      return this._borrowNextAvailableSession();
+      span.addEvent('Cache hit: has usable session');
+      const sessPromise = this._borrowNextAvailableSession();
```
Comment (surbhigarg92) [inline src/session-pool.ts:900]: why this change is required ?  to save results in const 

directly use `return this._borrowNextAvailableSession();`

File: test/session-pool.ts [hunk 8]
```diff
@@ -1349,6 +1365,84 @@ describe('SessionPool', () => {
...
+          spans.length,
+          1,
+          'Exactly 1 span should have been exported'
+        );
+        assert.strictEqual(
+          spans[0].name,
+          `${SPAN_NAMESPACE_PREFIX}.${topLevelSpanName}`,
+          'Expected only the top-level created span'
+        );
+        const span0 = spans[0];
```
Comment (surbhigarg92) [inline test/session-pool.ts:1426]: I don't see any assert on the annotations. 

Ideally we should add below tests 

1) Executing a query when no sessions are available

-> These will have "Acquiring Session", "Waiting for session to become available", "Creating Session" , "Acquired session" annotation

2) Executing a query when session is available"

 These will have "Acquiring Session", "Cache hit: has usable session" , "Acquired session" annotation

Comment (surbhigarg92) [reply test/session-pool.ts:1426]: Refer https://github.com/googleapis/java-spanner/blob/main/google-cloud-spanner/src/test/java/com/google/cloud/spanner/OpenTelemetrySpanTest.java

File: src/session-pool.ts [hunk 9]
```diff
@@ -645,12 +654,17 @@ export class SessionPool extends EventEmitter implements SessionPoolInterface {
       const elapsed = Date.now() - startTime;
 
       if (elapsed >= timeout!) {
+        setSpanErrorAndException(span, 'timed out');
         throw new GoogleError(errors.Timeout);
       }
```
[the remark this answers is not in the document]
Comment (odeke-em) [reply src/session-pool.ts:662]: Done, thanks!

File: test/session-pool.ts (hunk 8 above)
Comment (odeke-em) [reply test/session-pool.ts:1426]: There are assertions on the annotations for SessionPool,. Also in our discussions, we agreed that a follow-up PR this one or two shall be a PR to add tests against the Spanner Mock server which test the whole grouping hence we'll able to test out all the annotations.

File: src/session-pool.ts (hunk 7 above)
Comment (odeke-em) [reply src/session-pool.ts]: Sorry about that, it was a vestige from prior code that did something on the promise. Fixed, thanks!

File: src/session-pool.ts (hunk 4 above)
Comment (odeke-em) [reply src/session-pool.ts:1047]: Keeping that documentation in, then later on once we have traceWrap in, adding explicit tests that ping doesn't produce any spans.

File: src/session-pool.ts [hunk 10]
```diff
@@ -747,16 +768,19 @@ export class SessionPool extends EventEmitter implements SessionPoolInterface {
       } catch (e) {
         this._pending -= amount;
         this.emit('createError', e);
+        setSpanErrorAndException(span, e as Error);
         throw e;
       }
 
+      span.addEvent(`started adding ${amount} sessions to the inventory`);
```
[the remark this answers is not in the document]
Comment (odeke-em) [reply src/session-pool.ts]: Roger that!

File: src/session-pool.ts (hunk 3 above)
Comment (odeke-em) [reply src/session-pool.ts]: Done, thanks!

File: src/session-pool.ts [hunk 11]
```diff
@@ -931,11 +977,19 @@ export class SessionPool extends EventEmitter implements SessionPoolInterface {
       if (amount + this.size > this.options.max!) {
         amount = this.options.max! - this.size;
       }
+      span.addEvent('creating new session', {
```
[the remark this answers is not in the document]
Comment (odeke-em) [reply src/session-pool.ts]: Done, thanks!

File: src/session-pool.ts [hunk 12]
```diff
@@ -931,11 +977,19 @@ export class SessionPool extends EventEmitter implements SessionPoolInterface {
...
+        amount: amount,
+        min: min,
+      });
+
       if (amount > 0) {
         this._pending += amount;
         promises.push(
           new Promise((_, reject) => {
             this._pending -= amount;
+            span.addEvent('creating a new session');
```
[the remark this answers is not in the document]
Comment (odeke-em) [reply src/session-pool.ts]: Done, thanks!

File: src/database.ts [hunk 13]
```diff
@@ -821,7 +822,14 @@ class Database extends common.GrpcServiceObject {
         callback!(err as ServiceError, null, undefined);
         return;
       }
+
+      const span = getActiveOrNoopSpan();
+      span.addEvent('Creating BatchTransaction');
```
Comment (surbhigarg92) [inline src/database.ts:827]: Remove this event as I feel once we add the spans , there should be a span for "BatchTransaction" , so this information will be repeated in this case.

File: src/database.ts [hunk 14]
```diff
@@ -821,7 +822,14 @@ class Database extends common.GrpcServiceObject {
...
+
       const transaction = this.batchTransaction({session: session!}, options);
+      span.addEvent('BatchTransaction Creation Done', {
```
Comment (surbhigarg92) [inline src/database.ts:830]: This Span needs to move to the "transaction.begin" called below. Till the time transaction.begin is called, there will not be any transaction "id"

File: src/database.ts [hunk 15]
```diff
@@ -2951,6 +2962,9 @@ class Database extends common.GrpcServiceObject {
             if (session) {
               session.lastError = err as grpc.ServiceError;
             }
+            span.addEvent('No session available', {
+              'failed_session.id': session?.id,
```
Comment (surbhigarg92) [inline src/database.ts:2966]: keep the name same "session.id" instead of "failed_session.id"

File: src/database.ts [hunk 16]
```diff
@@ -2924,13 +2932,16 @@ class Database extends common.GrpcServiceObject {
...
     const proxyStream: Transform = through.obj();
+    const span = getActiveOrNoopSpan();
 
     this.pool_.getSession((err, session) => {
       if (err) {
         proxyStream.destroy(err);
         return;
       }
 
+      span.addEvent('Using Session', {'session.id': session?.id});
```
Comment (surbhigarg92) [inline src/database.ts:2943]: Can you add the same event for all the places, where we are calling `this.pool_.getSession` in this file ? 

Same goes for `No session available` event in case of "Sessionnotfounderror"

File: src/session-pool.ts [hunk 17]
```diff
@@ -630,8 +635,12 @@ export class SessionPool extends EventEmitter implements SessionPoolInterface {
    * @returns {Promise<Session>}
    */
   async _acquire(): Promise<Session> {
+    const span = getActiveOrNoopSpan();
     if (!this.isOpen) {
-      throw new GoogleError(errors.Closed);
+      span.addEvent('SessionPool is closed');
+      const err = new GoogleError(errors.Closed);
+      setSpanErrorAndException(span, err);
```
Comment (surbhigarg92) [inline src/session-pool.ts:642]: As per our discussion you are planning to create `tracewrapper` which will handle all the errors and event. 

Will this error be covered in that?

File: src/session-pool.ts [hunk 18]
```diff
@@ -642,18 +651,31 @@ export class SessionPool extends EventEmitter implements SessionPoolInterface {
     // wrapping this logic in a function to call recursively if the session
     // we end up with is already dead
     const getSession = async (): Promise<Session> => {
+      span.addEvent('Acquiring session');
       const elapsed = Date.now() - startTime;
 
       if (elapsed >= timeout!) {
+        setSpanErrorAndException(span, 'timed out');
```
Comment (surbhigarg92) [inline src/session-pool.ts:658]: nit: Change the message from 'timed out' to "Acquiring session timed out"

File: src/session-pool.ts [hunk 19]
```diff
@@ -744,9 +774,14 @@ export class SessionPool extends EventEmitter implements SessionPoolInterface {
         });
 
         amount -= sessions.length;
+        nReturned += sessions.length;
       } catch (e) {
         this._pending -= amount;
         this.emit('createError', e);
+        span.addEvent(
```
Comment (surbhigarg92) [inline src/session-pool.ts:781]: We can remove this span as we are setting exception below.

File: src/session-pool.ts [hunk 20]
```diff
@@ -758,6 +793,8 @@ export class SessionPool extends EventEmitter implements SessionPoolInterface {
         });
       });
     }
+
+    span.addEvent(`Requested for ${nRequested} sessions returned ${nReturned}`);
```
Comment (surbhigarg92) [inline src/session-pool.ts:797]: what is the need of creating `nRequested` and  `nReturned` , we can dircetly use `amount` and `sessions.length` respectively

File: src/session-pool.ts [hunk 21]
```diff
@@ -922,6 +974,7 @@ export class SessionPool extends EventEmitter implements SessionPoolInterface {
       let amount = this.options.incStep
         ? this.options.incStep
         : DEFAULTS.incStep!;
+
```
Comment (surbhigarg92) [inline src/session-pool.ts:977]: nit: remove empty line

File: src/session-pool.ts [hunk 22]
```diff
@@ -931,6 +984,7 @@ export class SessionPool extends EventEmitter implements SessionPoolInterface {
       if (amount + this.size > this.options.max!) {
         amount = this.options.max! - this.size;
       }
+
```
Comment (surbhigarg92) [inline src/session-pool.ts:987]: nit: remove empty line

File: src/session-pool.ts (hunk 4 above)
Comment (surbhigarg92) [reply src/session-pool.ts:1047]: understood.

File: src/session.ts [hunk 23]
```diff
@@ -480,7 +481,11 @@ export class Session extends common.GrpcServiceObject {
     options?: TimestampBounds,
     queryOptions?: google.spanner.v1.ExecuteSqlRequest.IQueryOptions
   ) {
-    return new Snapshot(this, options, queryOptions);
+    const span = getActiveOrNoopSpan();
+    span.addEvent('Creating Snapshot');
+    const snapshot = new Snapshot(this, options, queryOptions);
+    span.addEvent('Snapshot creation done', {id: snapshot?.id?.toString()});
```
Comment (surbhigarg92) [inline src/session.ts:487]: Again, the transaction is not getting created here. Add this line, only for the places where we are calling "transaction.begin"

File: src/session.ts [hunk 24]
```diff
@@ -480,7 +481,11 @@ export class Session extends common.GrpcServiceObject {
     options?: TimestampBounds,
     queryOptions?: google.spanner.v1.ExecuteSqlRequest.IQueryOptions
   ) {
-    return new Snapshot(this, options, queryOptions);
+    const span = getActiveOrNoopSpan();
+    span.addEvent('Creating Snapshot');
```
Comment (surbhigarg92) [inline src/session.ts:485]: remove this

File: src/session.ts (hunk 23 above)
Comment (surbhigarg92) [reply src/session.ts:487]: like this place 
https://github.com/googleapis/nodejs-spanner/blob/main/src/database.ts#L2033

and this https://github.com/googleapis/nodejs-spanner/blob/main/src/database.ts#L2773

I would recommend to add below event at https://github.com/googleapis/nodejs-spanner/blob/main/src/transaction.ts#L1419 . This method gets called from all the places whenever transaction gets created and assigns the id. This would mean you can remove the `BatchTransaction creation done` event as well.

`span.addEvent('Transaction creation done', {id: id?.toString()});`

File: src/session.ts [hunk 25]
```diff
@@ -497,7 +502,18 @@ export class Session extends common.GrpcServiceObject {
...
-    return new Transaction(this, undefined, queryOptions, requestOptions);
+    const span = getActiveOrNoopSpan();
+    span.addEvent('Creating Transaction');
+    const transaction = new Transaction(
+      this,
+      undefined,
+      queryOptions,
+      requestOptions
+    );
+    span.addEvent('Transaction Creation Done', {
```
Comment (surbhigarg92) [inline src/session.ts:513]: Same here, remove all the create transaction and Transaction Creation Done event and keep at only one place, recommended in above comment

File: src/transaction.ts [hunk 26]
```diff
@@ -1954,6 +1955,9 @@ export class Transaction extends Dml {
     optionsOrCallback?: CommitOptions | CallOptions | CommitCallback,
     cb?: CommitCallback
   ): void | Promise<CommitResponse> {
+    const span = getActiveOrNoopSpan();
+    span.addEvent('Starting Commit');
```
Comment (surbhigarg92) [inline src/transaction.ts:1959]: I know this is part of the scope, but as long as we add this method in "tracewrapper" , this event is not required, same goes for below event `Commit failed` , if the error gets sets, then its not requried.

File: src/transaction.ts [hunk 27]
```diff
@@ -2016,6 +2029,13 @@ export class Transaction extends Dml {
         }
         err = Transaction.decorateCommitError(err as ServiceError, mutations);
 
+        if (err) {
+          setSpanError(span, err);
+          span.addEvent('Commit failed');
```
Comment (surbhigarg92) [inline src/transaction.ts:2034]: no need for these events

File: test/spanner.ts [hunk 28]
```diff
@@ -4985,6 +4993,69 @@ describe('Spanner with mock server', () => {
...
+        return evtA.time < evtB.time;
+      });
+
+      const gotEventNames: string[] = [];
+      events.forEach(event => {
+        gotEventNames.push(event.name);
+      });
+
+      const wantEventNames = [
+        'Requesting 25 sessions',
```
Comment (surbhigarg92) [inline test/spanner.ts:5042]: These event names, doesn't look right? 

In case of success , shouldn't we have event like "Acquired session" and we shouldn't have 'Waiting for a session to become available' as it should do a cache hit.

File: package.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (surbhigarg92) [reply package.json]: We decided to remove this dependency as we have not released the last changes and its safe to remove.

File: package.json
Comment (surbhigarg92) [inline package.json:1]: We decided to remove this as we have not released the code changes which added this, "@opentelemetry/context-async-hooks"

File: src/session-pool.ts (hunk 20 above)
Comment (odeke-em) [reply src/session-pool.ts:797]: Sadly amount gets reduced on each iteration, so we can't use it, hence why we capture it upfront.

File: package.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (odeke-em) [reply package.json]: Roger that and thanks!

File: src/database.ts (hunk 13 above)
Comment (odeke-em) [reply src/database.ts:827]: Alright, thanks! Though in our scope document we needed to annotate when a transaction is started. Correct though that we shall have the span for BatchTransaction.

File: src/database.ts (hunk 14 above)
Comment (odeke-em) [reply src/database.ts:830]: Good point, thank you!

File: src/database.ts (hunk 15 above)
Comment (odeke-em) [reply src/database.ts:2966]: So "session.id" didn't make sense given the event "No session available", "failed_session.id" gave more prescription that it failed when we tried to retrieve using that sessionID, but no biggie I am making that change.

File: src/session-pool.ts (hunk 17 above)
Comment (odeke-em) [reply src/session-pool.ts:642]: Indeed, it'll handle setting the error. I suppose let me undo this setSpanErrorAndException here.

File: src/database.ts (hunk 16 above)
Comment (odeke-em) [reply src/database.ts:2958]: Roger that and done, thank you!

File: src/session-pool.ts (hunk 19 above)
Comment (odeke-em) [reply src/session-pool.ts:774]: I think we should keep it despite the exception because if for example it gets caught, that span should still have recorded the status at that terminal point.

File: test/spanner.ts (hunk 28 above)
Comment (odeke-em) [reply test/spanner.ts:5047]: So with how fragile the code in this file is, the creation of the Spanner object is precarious and is tightly coupled with the assertions of the test invocations and checks for suffixes with values like `*-instance-3`, so I made a TODO at the top of this `it` to later on refactor this code so that we can create our own Spanner object and check for all the events below. Those mentioned flakes were a huge delay for this review.

**Comment by odeke-em** [comment]
@surbhigarg92 thank you very much for the review! Kindly help me re-run the checks an also with a fresh set of review.

File: src/session-pool.ts [hunk 29]
```diff
@@ -771,6 +800,7 @@ export class SessionPool extends EventEmitter implements SessionPoolInterface {
    * @returns {Promise}
    */
   async _destroy(session: Session): Promise<void> {
+    const span = getActiveOrNoopSpan();
```
Comment (surbhigarg92) [inline src/session-pool.ts:803]: nit: this is not used

**Comment by odeke-em** [comment]
Yes indeed. Let's keep this one focused and smaller @surbhigarg92, thanks!

**Comment by odeke-em** [comment]
@willpoint this is the PR for us to base off. @surbhigarg92 kindly please take a look. In a follow-up PR I shall make the updates to span annotation tests for transactions but this change has transaction creation annotations.

File: src/session-pool.ts (hunk 29 above)
Comment (odeke-em) [reply src/session-pool.ts]: Done, thanks!

File: src/transaction.ts [hunk 30]
```diff
@@ -1966,12 +1973,24 @@ export class Transaction extends Dml {
...
 
+    const span = getActiveOrNoopSpan();
+
     if (this.id) {
       reqOpts.transactionId = this.id as Uint8Array;
     } else if (!this._useInRunner) {
       reqOpts.singleUseTransaction = this._options;
     } else {
-      this.begin().then(() => this.commit(options, callback), callback);
+      this.begin().then(
```
Comment (surbhigarg92) [inline src/transaction.ts:1983]: This is not required, as we are calling commit here again which will add `commit failed` event if error.

**Review by surbhigarg92 — APPROVED** [review]

**Comment by surbhigarg92** [comment]
@odeke-em I have approved the PR. But Please address the new comments before planning to merge.

File: src/database.ts [hunk 31]
```diff
@@ -3195,11 +3242,13 @@ class Database extends common.GrpcServiceObject {
         ? (optionsOrRunFn as RunTransactionOptions)
         : {};
 
+    let sessionId = '';
     const getSession = this.pool_.getSession.bind(this.pool_);
     // Loop to retry 'Session not found' errors.
     // (and yes, we like while (true) more than for (;;) here)
     // eslint-disable-next-line no-constant-condition
     while (true) {
+      const span = getActiveOrNoopSpan();
```
[the remark this answers is not in the document]
Comment (odeke-em) [reply src/database.ts:3251]: I had it in the loop to get the latest current active span in case the code down below invoked this method recursively. Removed though, thank you!

File: src/database.ts [hunk 32]
```diff
@@ -3424,14 +3483,20 @@ class Database extends common.GrpcServiceObject {
         ? (optionsOrCallback as CallOptions)
         : {};
     this.pool_.getSession((err, session?, transaction?) => {
+      const span = getActiveOrNoopSpan();
       if (err && isSessionNotFoundError(err as grpc.ServiceError)) {
+        const span = getActiveOrNoopSpan();
```
[the remark this answers is not in the document]
Comment (odeke-em) [reply src/database.ts:3488]: Definitely left over from a prior change, done thanks!

File: src/session-pool.ts [hunk 33]
```diff
@@ -865,21 +894,28 @@ export class SessionPool extends EventEmitter implements SessionPoolInterface {
...
     let removeOnceCloseListener: Function;
     let removeListener: Function;
 
     // Wait for a session to become available.
+    span.addEvent('Waiting for a session to become available');
     const availableEvent = 'session-available';
     const promises = [
       new Promise((_, reject) => {
-        const onceCloseListener = () => reject(new GoogleError(errors.Closed));
+        const onceCloseListener = () => {
```
[the remark this answers is not in the document]
Comment (odeke-em) [reply src/session-pool.ts:915]: Oops, it was left over from a prior change, thanks and done!

File: src/transaction.ts [hunk 34]
```diff
@@ -413,6 +414,9 @@ export class Snapshot extends EventEmitter {
       options,
     };
 
+    const span = getActiveOrNoopSpan();
+    span.addEvent('Creating Transaction');
```
[the remark this answers is not in the document]
Comment (odeke-em) [reply src/transaction.ts:418]: Done, thanks!

File: src/transaction.ts (hunk 30 above)
Comment (odeke-em) [reply src/transaction.ts]: Done, thanks!
